Manage Experiments from MLflow UI (Create/Rename/Delete Experiments)

ggliem opened the pull request:
## What changes are proposed in this pull request?

Manage Experiments from the UI:
- Added an option to create experiments to the UI
- Added an option to rename experiments to the UI
- Added an option to delete experiments to the UI

Instead of using the CLI, users can now manage experiments via the UI more easily.
These features (among others) have also been requested in https://github.com/mlflow/mlflow/issues/1028.

The following gif demonstrates the change:

![mlflow_manage_experiments_v2](https://user-images.githubusercontent.com/15968047/73715228-1fd9ea80-46c8-11ea-9736-c79b9e25f3de.gif)

## How is this patch tested?

- Manual testing
- Added basic unit tests to `ExperimentListView.test.js` and related forms

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Enable users to create, delete and rename experiments from the UI.

### What component(s) does this PR affect?

- [X] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [ ] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

ggliem committed:
Add delete and update experiment endpoints to service

Added delete and update experiment endpoints to existing
MlflowService implementation. These operations were already supported by
MLflows REST API. For more information, see:
https://www.mlflow.org/docs/latest/rest-api.html
Added corresponding AJAX calls.

ggliem committed:
Add edit and delete button to experiment list

Added edit and delete button to experiment list view.
The buttons do not have any functionality associated with them yet.

ggliem committed:
Add delete experiment functionality

Added functionality to delete experiments.
When clicking on the delete button in the experiment list,
 confirmation modal is opened.

ggliem committed:
Refactor and generalize modal to rename runs

Updated modal to rename runs such that it is more generic.
In particular, updated the displayed form.
As such, this component can be used for renaming experiments.

ggliem committed:
Update delete experiment modal

Updated delete experiment modal such that the passed experimentId is
a number instead of a string. Updated ExperimentListView accordingly.

ggliem committed:
Add rename experiment functionality

Added functionality to rename experiments.
When clicking on the edit button in the experiment list,
 a modal is opened asking for a new experiment name.

ggliem committed:
Fix rename experiment functionality

Fixed a minor issue which caused the experiment name to not be updated.

ggliem committed:
Refresh experiment list after deletion

Updated experiment deletion functionality such that
the shown experiment list is refreshed/updated after
performing the deletion.
To do so, the behavior of experimentById reducer had
to be modified.
The previous implementation kept deleted experiments in the state
until the page was refreshed. This could also be observed when
deleting an experiment via the CLI since
the deleted experiment would remain in the UI until refreshing.

ggliem committed:
Fix existing unit tests

ggliem committed:
Add unit test

Added unit tests testing that the delete and rename
experiment modals are opened after clicking the
respective links in the experiment list.

ggliem committed:
Update styles of rename experiment modal

ggliem committed:
Disable delete experiment link for active experiment

Updated the experiment list in such a way that
the "Delete Experiment" option is disabled
for the currently active/selected experiment.

ggliem committed:
Fix linting issue

Zangr commented:
Thank you @ggliem for adding the delete/rename feature! The component layout LGTM in this PR. We are currently migrating to **antd** as our new design language and component library, so it'll be great to use antd in new components instead of other 3rd party library like ReactModal or Formik. I'd suggest using antd modal and form as shown in [this example](showDeleteExperimentModal) or [RegisterModelButton](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/model-registry/components/RegisterModelButton.js). It will be even better if you can point the rename modal in run page to the new modal as well. We'll be glad to review your changes :)

harupy commented on mlflow/server/js/src/components/ExperimentListView.js:
```diff
@@ -57,19 +118,36 @@ export class ExperimentListView extends Component {
-                    {e.getName()}
-                  </div>
-                </Link>
+                    <Icon type="delete" />
```
@ggliem 

The image below is how the delete-tag button looks like on the run view page. I think the delete-experiment button should use the same icon.

<img width="423" alt="Screen Shot 2020-01-28 at 16 56 05" src="https://user-images.githubusercontent.com/17039389/73245461-9c178e00-41ef-11ea-8190-7ce9cf768d78.png">


Code

https://github.com/mlflow/mlflow/blob/690c7e933f4c1007f477fa2b240631a9f8f515cf/mlflow/server/js/src/components/tables/EditableFormTable.js#L119-L121

harupy replied:
btw this feature looks awesome!

ggliem replied:
Sure, I can change that (I actually used the same implementation in the beginning)

ggliem committed:
Migrate renaming modals to antd

Based on the following input of the created PR:
https://github.com/mlflow/mlflow/pull/2348#issuecomment-579087684
Updated rename modals to use antd modals and forms instead of ReactModal and Formik.

ggliem commented:
> Thank you @ggliem for adding the delete/rename feature! The component layout LGTM in this PR. We are currently migrating to **antd** as our new design language and component library, so we are trying to use antd components if possible in new changes. For this PR I'd suggest using antd Modal and Form as shown in [this example](showDeleteExperimentModal) or [RegisterModelButton](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/model-registry/components/RegisterModelButton.js) instead of ReactModal + Formik. It will be even better if you can point the rename modal in the run page to the new antd modal as well. We'll be glad to review your changes :)

Thank you @Zangr for your feedback. I started migrating the modals to antd. I changed both the "RenameExperimentModal" as well as the "RenameRunModal" to use antd's Modal&Form instead of ReactModal&Formik in the recent commit. I will update the "DeleteExperimentModal" accordingly. While migrating the "Rename... modals" to antd I noticed the following things:

1. It seems like setting the focus on the input field does not work properly when reopening the modal. I tried calling `.focus()` on the selected reference (Input) but it did not have the desired effect. I also found some articles online stating that that does not work. As such, the user has to click on the input field first.  I noticed that that RegisterModelModal has the same 'issue'. Do you know a workaround for that?
2. Compared to ReactModal, the modal seems to not have an 'Enter' event (https://github.com/ant-design/ant-design/issues/11088) (by design). This also holds for the RegisterModelModal.
3. I used the `getFieldDecorator` in the form. It seems like I can't change the initialValue unless I unmount `Form.Item`. Compared to the previous implementation, the previous run/experiment name is thus not displayed in the input field when opening the modal. Could you please clarify whether that's ok?
![rename_experiment](https://user-images.githubusercontent.com/15968047/73247051-ed2e7000-4164-11ea-9ebb-d3ff2e65bb75.png)

harupy commented on mlflow/server/js/src/components/ExperimentListView.css:
```diff
@@ -48,3 +48,7 @@
+
+.rename-experiment-form {
+  width: 480px;
+}
```
nit: no newline at end of file.

Zangr commented:
@ggliem Thank you for the refactoring! To solve the 3 issues you mentioned, we can leverage **Ref** and some imperative ops in **componentDidUpdate**. Here is a [sample modal implementation](https://stackblitz.com/edit/react-t6ukr9?file=index.js) with solutions to all you asks.
![ref-didupdate-fix](https://user-images.githubusercontent.com/6811562/73324881-59f74800-4201-11ea-904a-da2de31c4cca.gif)

ggliem commented:
> @ggliem Thank you for the refactoring! To solve the 3 issues you mentioned, we can leverage **Ref** and some imperative ops in **componentDidUpdate**.
> 
Nice, thanks for the input! I will go ahead and change it.

ggliem committed:
Add focus and initialValue to input in rename modals

Updated RenameFormView used in rename modals.
The input of the modal will be focused and selected after (re)opening the modal.
The input of the modal will contain the respective name of the experiment/run as the initial value
after re(opening) the modal.

ggliem committed:
Create generic input modal & refactor rename modals

Refactored rename modals.
Since the RenameExperimentModal and RenameRunModal had a lot in common,
I put most code in a generic modal component.
This generic modal component is used by the rename modals.

ggliem committed:
Migrate ConfirmModal to antd

Based on the following input of the created PR:
https://github.com/mlflow/mlflow/pull/2348#issuecomment-579087684
Updated the generic ConfirmModal to use antd modals.

ggliem committed:
Update "Delete" icon

ggliem commented:
@Zangr I implemented the changes & ended up refactoring the "Rename" modals even further by putting  the shared code in a generic component. I also migrated the ConfirmModal to antd.

I noticed that when calling the 'Update Experiment' API (`/experiments/update`) there is no check whether the experiment name already exists. In contrast, the backend service throws an error when trying to create a new experiment with the same name.
Since having experiments with the same name could cause some issues (e.g. when calling `mlflow.set_experiment(experiment_name)`), I plan to change the service implementation later. Could you please confirm that this is in fact a bug?

ggliem commented:
@Zangr One more question: I noticed that the CI pipeline  failed while executing the "Small" Python (Build: `7746`) or "Large" Python (Build: `7709`) step. Could you please provide some insight whether there is a way to fix that?

ggliem committed:
Fix error handling of generic modal

Fixed error handling of generic modal.
Before this change, the error modal was not correctly displayed.

Zangr reviewed (commented):
Manually tested the code change, works pretty well. I left some comments mostly related to naming/usage convention and discussion wrapper component extraction.

Zangr commented:
cc @smurching for the question on python test failure

ggliem commented on mlflow/server/js/src/components/ExperimentListView.js:
```diff
@@ -57,19 +118,36 @@ export class ExperimentListView extends Component {
+                  <Link
+                    style={{ textDecoration: 'none', color: 'unset', width: '80%'}}
+                    to={Routes.getExperimentPageRoute(e.getExperimentId())}
+                    onClick={active ? ev => ev.preventDefault() : ev => ev}
```
I generally agree but figured that the following problem had to be solved:
If a user is e.g. reviewing experiment `2`, the URL will point to `/experiments/2`. After deleting the experiment the user would be redirected to the "Something went wrong..." screen since the UI is not handling the display of experiments which are associated with the lifecycle state "Deleted".

In the beginning, I actually looked into different solutions for the issue but I did not want to make too many assumptions about the expected behavior without getting feedback. I think the following solutions make sense and are quite easy to implement:
1. Make use of an imperative API (`history.push()`) after the deletion. This would require to wrap another component into the `withRouter` HOC.
2. Modifying the following part of the `HomeView` component and generally checking whether the experiment is active, followed by possible rerouting. However, this would change the general behavior of the application.
```js
  if (ownProps.experimentId === undefined) {
    const firstExp = getFirstActiveExperiment(getExperiments(state));
    if (firstExp) {
      return { experimentId: parseInt(firstExp.experiment_id, 10) };
    }
  }
  return {};
```

If you could provide me with the expected behavior I could change it accordingly.

I tried to mention that in the description of the PR:

> The "Delete Experiment" option is disabled for the currently active/selected experiment. Alternatively, the expected behavior needs to be defined for this case.

ggliem commented on mlflow/server/js/src/components/modals/GenericInputModal.js:
```diff
@@ -0,0 +1,108 @@
+        const newName = values[NEW_NAME_FIELD];
+        // call handleSubmit from parent component, pass input field value
+        // handleSubmit is expected to return a promise
+        this.props.handleSubmit(newName)
```
1. True, I will change that.
2. I could change that as well if we want to make the component even more generic.

However, I am not so sure about whether we should rather create a general form wrapper instead. I could definitely see arguments for doing that. However, I think that the operations like submitting, closing & handling submitting failures are still happening on the modal level. I am not sure whether changing it to a generic form wrapper instead would result in a more simple solution.

ggliem replied:
Changed both 1 and 2. I haven't changed it to a form wrapper yet though (see comment above)

Zangr replied:
I think it's OK to implement each Modal+Form separately in this PR. Some duplicated form handling appear at the modal layer for each implementation but the component name and code path will be very readable. We can work on a more general form wrapper as a follow up.

ggliem replied:
> I think it's OK to implement each Modal+Form separately in this PR.

Could you please specify this - does this refer to the separate modals for Delete&Rename? Can this conversation be resolved?

ggliem commented on mlflow/server/js/src/components/modals/InputFormView.js:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Component that renders a form for updating a run's or experiment's name.
+ */
+class InputFormViewComponent extends Component {
```
That was actually the first name that I chose (`RenameModal` respectively). However, I thought that the component (as well as the "generic" modal) could be used for a `Create Experiment` use case as well. If you could provide me with a mockup that includes some kind of `Create` button, I could add it to the PR. The "create experiment from UI" has also been requested by the community before.

gioa replied:
Thank you for adding the rename and delete features! Looks great to me.

For the new experiment button, see a mock below. I use a 13px Font awesome **fa-plus** with the following styles `width: 24px;
  height: 24px;
  color: #333333;
  background-color: #f5f5f5;
  border-color: #cccccc;
  border-radius: 4px;
  margin-right: 12px;
`  

![MLFlow_NewExperiment](https://user-images.githubusercontent.com/1960721/73681541-67825700-4673-11ea-93fb-8e94c70d0483.png)

ggliem replied:
Thanks for providing the mockup. I added the functionality and updated the PR description.
CC @Zangr for the code review.

gioa replied:
Thank you for adding this so quickly! Btw shall we direct users to the newly created experiment page after the creation dialog? So they don't need to manually find the new experiment in the list.

ggliem replied:
Sure, we could do that. I also thought about it while implementing the feature but wasn't sure about it since no runs will be shown until executing a run & refreshing the page. 
Would you like me to change it?

gioa replied:
Yes! I assume the next natural step for users is to copy the experiment ID to log runs. So it would be nice to show the new experiment.

ggliem replied:
I totally agree! The redirect should work now.

ggliem commented:
Thanks @Zangr for the feedback. I replied to some of your comments where I had some follow-up questions/remarks.

ggliem committed:
Align with naming conventions & use wrapped dispatch functions explicitly

Incorporated feedback from PR.
Aligned with naming conventions (callbacks, variables...)
Moved follow-up calls from deleteExperimentApi and deleteRunApi up to be explicitly invoked
as a follow up in the then blocks.
Updated code to use wrapped dispatch functions explicitly.

ggliem committed:
Add basic rendering test for InputFormView

ggliem commented on mlflow/server/js/src/components/modals/InputFormView.js:
```diff
@@ -0,0 +1,68 @@
+import React, { Component } from 'react';
```
Added a basic rendering test. Could you please double check whether that was what you expected?

Zangr replied:
Yes, this looks good. I added some test naming nit that we've been following in other components.

ggliem committed:
Update GenericInputModal

Refactored GenericInputModal, making the component more generic.
1. A form component can now be passed to the modal as a render prop
2. The entire values object (from the form) is now passed to the submit callback.
The caller can thus decide what to do with the values object.

Zangr commented on mlflow/server/js/src/components/modals/InputFormView.test.js:
```diff
@@ -0,0 +1,18 @@
+    form: { getFieldDecorator: jest.fn(opts => c => c) },
+  };
+
+  it('renders with props', () => {
```
"should render with minimal props without exploding"

Zangr commented on mlflow/server/js/src/components/modals/InputFormView.test.js:
```diff
@@ -0,0 +1,18 @@
+import { InputFormView } from './InputFormView';
+
+describe('Render test', () => {
+  const defaultProps = {
```
```suggestion
  const minimalProps = {
```

ggliem committed:
Align with naming conventions inside basic rendering test

ggliem committed:
Rename InputFormView to RenameForm

ggliem committed:
Allow users to delete the current experiment

Updated the experiment deletion behavior.
Users can now delete the currently selected experiment.
As a consequence, they are re-routed to the root which then picks the next active experiment to show.

ggliem committed:
Add "Create Experiment" functionality

Added functionality to create experiments from the UI.
Added "Create Experiment" button based on mockup in PR #2348.
Added modal and form that handles experiment creation.
Added basic unit tests.

ggliem committed:
Route user to new experiment page after creation

Updated the behavior when creating an experiment.
After creation, the user will be redirected to the newly created experiment page.
As such, they don not need to manually find the new experiment in the list

gioa commented:
@ggliem Thank you for making the change! I got an error banner flashing by when creating a new experiment that already exists in the trash. Can we handle and display the error on the error dialog instead? So users can see what went wrong in one place. Thanks!

<img width="1101" alt="Error" src="https://user-images.githubusercontent.com/1960721/73868693-c673d780-47fd-11ea-92e8-62a059b12dab.png">

ggliem commented:
Good catch, thanks for pointing that out! I totally agree, let's try to make it consistent. You suggested to display the error on the error dialog. However, I'd like to clarify the behavior again before implementing the change.

Looking at the error handling of the [RegisterModelModal](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/model-registry/components/RegisterModelButton.js) I noticed that the error is displayed in an Antd-Message (the one displayed on the top of the screen in your screenshot), without closing the actual modal nor displaying the error dialog. The actual error is displayed in the message (e.g. "Internal Server Error").
On the contrary, the modals [DeleteRunModal](https://github.com/mlflow/mlflow/blob/f0a05237abc8b47ff8094064b28255dedca0097d/mlflow/server/js/src/components/modals/DeleteRunModal.js) or [RestoreRunModal](https://github.com/mlflow/mlflow/blob/241d43cd656f2a4619563f77ddf7354799427ff6/mlflow/server/js/src/components/modals/RestoreRunModal.js) display this kind of error dialog instead, while using a generic error message (e.g. 'While restoring an experiment run, an error occurred.').

It seems like I ended up mixing these two options. If an error occurs while creating/deleting/renaming an experiment, should we:
1) Close the current modal & show the error modal or
2) Keep the current modal open & display this kind of message at the top?

I am leaning towards option 2) since the user might have the chance to directly solve the issue (e.g. modifying the name) - what do you think?

gioa commented:
@ggliem Thank you for looking at all these different implementations and suggesting a better solution! I agree the option 2 is better given the user can update the name directly there. I hope the message does not dismiss itself before users finish reading. 

IMO an even better option is to display the error directly on the create dialog, similar to RenameRunFormView. So users can see the error and update the experiment's name right there. What do you think?

CC @Zangr Shall we look at all the different error handlings and put together a guideline for when to use what components?

ggliem commented:
@gioa The time the message is shown is configurable, so we could pick an appropriate time such that it doesn't dismiss itself before users finish reading.

I am not sure about the option to display it directly on the dialog. I personally like the message option better. I think a general guideline makes sense such that the errors are handled in a consistent way.
Does it make sense to have different guidelines for Create/Rename and Delete? I think for create/rename it definitely makes the most sense to have either a message or display the error directly in the modal but I am not sure whether the same holds for the "delete" modals. What do you think?

gioa commented:
Agree the create/rename and the delete can be handled differently. Create and rename's errors can be corrected by the user input while delete is often due to a system error - the user cannot do much about it but retrying. 

Just curious why you prefer the message option rather than display it directly on a dialog. Would love to learn what users think. Also for comparison, I mocked the two options as below. 

<img width="768" alt="ErrorMessage_2options" src="https://user-images.githubusercontent.com/1960721/73987160-5431ee80-48f4-11ea-8e6f-5cc733b40662.png">

ggliem commented:
@gioa Should we agree on handling the deletion errors in such a way that we display this kind of generic error modal?

Thanks for providing the side-by-side comparison. Seeing the direct comparison, I like the mockup in which the error is directly displayed on the dialog. However, one of the reasons I initially did not like the idea of showing it directly in the dialog too much is that the returned error could also be rather technical, e.g. an "Internal Server Error" which does not have to be related to the input field. Hence, I don't know whether the input field should be highlighted. This is how it looked like in the RenameRunFormView which I personally did not like too much:

![mlflow_rename_run_error](https://user-images.githubusercontent.com/15968047/74001504-bc96c500-4920-11ea-94fa-829604fce57c.png)

Another thing we could do is handling the "name duplication" error via form validators, similar to the validator in the [RegisterModelForm](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/model-registry/components/RegisterModelForm.js). We could check for name duplication (while renaming or creating experiments) without calling the API. However, the UI currently only keeps track of the active experiments. To make the behavior consistent while using the validators we would probably also have to keep track of the deleted experiments. We could do so by modifying the `listExperiments` call a bit but I am not sure whether that's intended (cc @Zangr)? The validator checks correspond to your mockup. Errors like the "Internal Server Error" could then be displayed in a message.

Zangr commented:
> To make the behavior consistent while using the validators we would probably also have to keep track of the deleted experiments. We could do so by modifying the `listExperiments` call a bit but I am not sure whether that's intended (cc @Zangr)?

Yes, to check the experiment on the client side we can do an on-demand **list_experiments** call with an extra query parameter **?viewType=ALL** and leverage the [async validator](https://github.com/yiminghe/async-validator) built-in with antd form to do the validation against current name in the form input.

Zangr commented:
> CC @Zangr Shall we look at all the different error handlings and put together a guideline for when to use what components?

Yes, we should. Overall my feeling is that errors from user actions fit more naturally with the pop ups while 4xx/5xx errors from accessing the main resource of the page during initial visiting can have a dedicated 4xx/5xx page for them.

ggliem commented:
> @ggliem Yes, to check the experiment on the client side we can do an on-demand **list_experiments** call with an extra query parameter **?viewType=ALL** and leverage the [async validator](https://github.com/yiminghe/async-validator) built-in with antd form to do the validation against current name in the form input.

Do you think it makes more sense to call `listExperiments` when rendering the modals (Rename/Create)  (i.e. on-demand) or keep the deleted experiments in the global store as well? If doing on-demand calls I would add the query parameter `?viewType=DELETED_ONLY` since the active ones are part of the Redux store anyways.

gioa commented:
@ggliem Thank you for explaining! Your reasons are quite legit and I agree that the internal server error does not relate to the input, displaying it in a message on the top makes more sense in terms of consistency (I remember seeing internal server error on a message). 

I like your suggestion of using the form validator to check the input and display an input error directly on the modal. For other system errors like internal server error, we can display in a message. I also assume deletion errors are not input errors, so we can display them in a message as well.

Zangr commented:
> @Zangr Do you think it makes more sense to call `listExperiments` when rendering the modals (Rename/Create) (i.e. on-demand) or keep the deleted experiments in the global store as well? If doing on-demand calls I would add the query parameter `?viewType=DELETED_ONLY` since the active ones are part of the Redux store anyways.

Actually, I think a more efficient way to do on-demand validation is to call [/mlflow/experiments/get-by-name](https://www.mlflow.org/docs/latest/rest-api.html#get-experiment-by-name), that way we don't need to worry about over fetching.

ggliem commented:
> Actually, I think a more efficient way to do on-demand validation is to call [/mlflow/experiments/get-by-name](https://www.mlflow.org/docs/latest/rest-api.html#get-experiment-by-name), that way we don't need to worry about over fetching.

@Zangr Are you suggesting to call `/mlflow/experiments/get-by-name` as part of the async-validator?  I wanted to clarify since doing so would result in an API call with every new input (each character). Please correct me if I'm mistaken.
Or are you suggesting to make the call once the user tries to submit? If so, I feel like we could directly call `/create`/`/update`.

Zangr commented:
> @Zangr Are you suggesting to call `/mlflow/experiments/get-by-name` as part of the async-validator? I wanted to clarify since doing so would result in an API call with every new input (each character). Please correct me if I'm mistaken.
> Or are you suggesting to make the call once the user tries to submit? If so, I feel like we could directly call `/create`/`/update`.

Good question. To prevent excessive calls, I think we can debounce the triggering of the validator.

ggliem committed:
Add getExperimentByName endpoint to service

Added getExperimentByName to MLflowService.
https://www.mlflow.org/docs/latest/rest-api.html#get-experiment-by-name

ggliem committed:
Update error handling & validation checks

Unified modal error handling based on the discussion in PR #2348.
Updated Rename/Create experiment modals to check whether the experiment name is already in use.
Leveraged async-validator to do the validation against the
current experiment name in the form input.
Added debouncing to the validation trigger to prevent excessive calls.
Other errors, e.g. Internal Server Errors, are now displayed in a message instead of a error dialog.

ggliem commented:
@Zangr @gioa I updated the error handling. The form validator is used to check for name duplication. Other system errors like internal server errors are now displayed in a message, without showing the error modal. The same holds for deletion errors. Feel free to review. I set the debounce time to 400ms and it seemed like a good balance to me, however, please let me know if you think we should update the time.

The error modal is still used in `DeleteRunModal` and `RestoreRunModal` but I think this could be changed in a follow-up PR that updates/unifies the error handling throughout the application.

H4dr1en commented:
Thanks for the feature @ggliem! Looks awesome! One question that comes into my mind:
Would it be possible when creating a experiment to enter not only name of the experiment, but also `artifac_location`, so that it matches the `create_experiment` signature?

That would be important in my case, where I would like to specify an s3 bucket location for artifacts. 👍

Zangr commented on mlflow/server/js/package.json:
```diff
@@ -17,6 +17,7 @@
     "json-bigint": "databricks/json-bigint#a1defaf9cd8dd749f0fd4d5f83a22cd846789658",
     "leaflet": "^1.5.1",
     "lodash": "^4.17.11",
+    "lodash.debounce": "^4.0.8",
```
Do we need a new package for this? Will `_.debouce` do the same?

ggliem replied:
Generally speaking, `_.debounce` will do the same. However, as far as I know, lodash itself can be pretty heavy. That's why I imported a specific module instead of importing the whole lodash library. As far as I know `import _ from 'lodash'` will not tree-shake. I am not 100% sure about the newer versions, so please let me know if my assumptions are not correct. Would you like me to use `_.debounce` instead?

Zangr commented on mlflow/server/js/src/components/modals/validation.js:
```diff
@@ -0,0 +1,25 @@
+import { wrapDeferred } from '../../Actions';
+import { MlflowService } from '../../sdk/MlflowService';
+
+export const getExperimentNameValidator = (experimentNames) => {
```
Basic tests for the validator?

Zangr commented on mlflow/server/js/src/components/modals/GenericInputModal.js:
```diff
@@ -91,9 +86,5 @@ export class GenericInputModal extends Component {
-};
-
-export default connect(undefined, mapDispatchToProps)(GenericInputModal);
+export default GenericInputModal;
```
Already exported above, prefer using explicit export/import.

Zangr reviewed (approved):
Generally LGTM! I added some minor comments.

Also, let's disable the "Save" button when there is an error present? cc @gioa 
![Screen Shot 2020-02-14 at 8 51 49 AM](https://user-images.githubusercontent.com/6811562/74551176-91efd200-4f07-11ea-89b7-47247647d875.png)

Zangr commented on mlflow/server/js/src/components/modals/CreateExperimentModal.js:
```diff
@@ -0,0 +1,80 @@
+
+    const inputComponent = <CreateExperimentForm
+      visible={isOpen}
+      validator={debounce(experimentNameValidator, 400)}
```
This creates a new instance of validator function per renders no matter it's used or not.  We can bind it outside the render and call it deboundedExperienceNameValidator.

```
  debouncedExperimentNameValidator = _.debounce(
    getExperimentNameValidator(() => this.props.experimentNames),
    400,
  );
  
  export const getExperimentNameValidator = (getExistingExperimentNames) => {

    ...

    } else if (getExistingExperimentNames().includes(value)) {

    ...
```

ggliem replied:
Good point, I will update this.

ggliem committed:
Update debounced validator usage & use explicit exports

Updated debounced validator. Before, a new instance of the validator function was created per render.
Bound debounce outside the render.
Updated some components to use explicit exports.

ggliem committed:
Add unit test for experimentNameValidator

ggliem committed:
Add artifact location to CreateExperimentModal

Updated CreateExperimentModal. A user can now specify an artifact location.
Creating an experiment now matches the overall create_experiment signature

ggliem commented:
> Also, let's disable the "Save" button when there is an error present? cc @gioa

In terms of consistency, we would probably also have to update the [RegisterModelModal](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/model-registry/components/RegisterModelButton.js).
I'd also like to point out that this is aligned with Antd`s default behavior:
https://ant.design/components/form/

ggliem commented:
> Would it be possible when creating a experiment to enter not only name of the experiment, but also `artifact_location`, so that it matches the `create_experiment` signature?

@H4dr1en Thanks for pointing it out, it should be fixed with my recent commit.

ggliem committed:
Update named exports & update lodash import

Zangr commented:
> In terms of consistency, we would probably also have to update the [RegisterModelModal](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/model-registry/components/RegisterModelButton.js).
> I'd also like to point out that this is aligned with Antd`s default behavior:
> https://ant.design/components/form/

I see. I think it's fine to leave as-is.

Zangr commented:
@ggliem , did some manual test, mostly good except I'm getting an NPE while creating experiment from UI. Could you pleae take a look?

![npe-create](https://user-images.githubusercontent.com/6811562/74597883-7d046300-501b-11ea-98ac-ed04992c24cf.gif)

ggliem commented:
> @ggliem , did some manual test, mostly good except I'm getting an NPE while creating experiment from UI. Could you pleae take a look?
> 

Unfortunately, I can't reproduce the error. I tried creating multiple experiments with different configurations (e.g. with & without artifact location) but I couldn't get it to break locally. Could you please provide some more detail? Did you experience this error before? Does the creation of the experiment work and it's a issue with the routing or is no experiment created at all? I tried throwing an error in the backend, however, an error message is shown & I am not re-routed. Does the ID in the updated URL match the experiment ID? 

![mlflow_create_exp](https://user-images.githubusercontent.com/15968047/74598302-1df71c00-5024-11ea-85c0-c2466cd2ea98.gif)

Zangr commented:
> Unfortunately, I can't reproduce the error. I tried creating multiple experiments with different configurations (e.g. with & without artifact location) but I couldn't get it to break locally. Could you please provide some more detail? Did you experience this error before? Does the creation of the experiment work and it's a issue with the routing or is no experiment created at all? I tried throwing an error in the backend, however, an error message is shown & I am not re-routed. Does the ID in the updated URL match the experiment ID?

I think it's more easily reproducible when there are larger experiments.
Could you please try the following:
- Rebasing your change on top the latest master
- rm -rf mlruns (in mlflow project root folder)
- python tests/generate_ui_tetest_data.py --large (this will create a new mlruns folder with large test cases)
- mlflow server
- in another console, cd mlflow/server/js
- npm start

It's likely caused by the concurrent requests here:
 `Promise.all([createExperimentPromise, listExperimentsPromise])`
I'd recommend putting list experiment in a then block after create experiment, that way we can make sure list experiment gets up-to-date information. 

```
    return this.props
      .createExperimentApi(experimentName, artifactLocation, createRequestId)
      .then(() => this.props.listExperimentsApi(createRequestId))
      .then(([{ value }]) => {
        if (value && value.experiment_id) {
          // redirect user to newly created experiment page
          this.props.history.push(Routes.getExperimentPageRoute(value.experiment_id));
        }
      });
```
Also, it would be nice to add a reducer entry for  CREATE_EXPERIMENT_API as well.

Zangr reviewed (changes requested):
> Unfortunately, I can't reproduce the error. I tried creating multiple experiments with different configurations (e.g. with & without artifact location) but I couldn't get it to break locally. Could you please provide some more detail? Did you experience this error before? Does the creation of the experiment work and it's a issue with the routing or is no experiment created at all? I tried throwing an error in the backend, however, an error message is shown & I am not re-routed. Does the ID in the updated URL match the experiment ID?

I think it's more easily reproducible when there are larger experiments.
Could you please try the following:
- Rebasing your change on top the latest master
- rm -rf mlruns (in mlflow project root folder)
- python tests/generate_ui_tetest_data.py --large (this will create a new mlruns folder with large test cases)
- mlflow server
- in another console, cd mlflow/server/js
- npm start

It's likely caused by the concurrent requests here:
 `Promise.all([createExperimentPromise, listExperimentsPromise])`
I'd recommend putting list experiment in a then block after create experiment, that way we can make sure list experiment gets up-to-date information. 

```
  createRequestId = getUUID();
  listRequestId = getUUID();

  handleCreateExperiment = (values) => {
    // get values of input fields
    const experimentName = values[EXP_NAME_FIELD];
    const artifactLocation = values[ARTIFACT_LOCATION];

    return this.props
      .createExperimentApi(experimentName, artifactLocation, this.createRequestId)
      .then(() => this.props.listExperimentsApi(this.listRequestId))
      .then(([{ value }]) => {
        if (value && value.experiment_id) {
          // redirect user to newly created experiment page
          this.props.history.push(Routes.getExperimentPageRoute(value.experiment_id));
        }
      });
  };
```
Also, it would be nice to add a reducer entry for  CREATE_EXPERIMENT_API as well.

Zangr commented on mlflow/server/js/src/Actions.js:
```diff
@@ -45,6 +45,48 @@ export const getExperimentApi = (experimentId, id = getUUID()) => {
+        new_name: newExperimentName,
+      }),
+      meta: { id: getUUID() },
+    }).then(() => dispatch(getExperimentApi(experimentId, id)));
```
Similar to create an experiment, let's move this out to the rename model layer to keep this operation do what its called.

ggliem committed:
Update updateExperimentApi

Moved follow-up getExperimentApi call to the rename modal layer.

ggliem committed:
Fix NEP error after experiment creation

Updated the experiment creation behavior.
Before, the sequence of the performed calls after a user created an experiment was not guaranteed.
As such, the experiment list was sometimes not refreshed before the rerouting happened which led to an error in the UI.

ggliem committed:
Update formatting

ggliem commented:
> I think it's more easily reproducible when there are larger experiments. [...]
> It's likely caused by the concurrent requests here:
> `Promise.all([createExperimentPromise, listExperimentsPromise])`
> I'd recommend putting list experiment in a then block after create experiment, that way we can make sure list experiment gets up-to-date information.

I could not reproduce the issue locally, however, it was likely caused by the mentioned issue since that did not guarantee the order in which the promises were resolved. That being said, I do not think that the suggested code works since the state of `createExperimentApi` promise has to be shared through the promise chain. I pushed a change - could you please check whether you still get the error?

> Also, it would be nice to add a reducer entry for CREATE_EXPERIMENT_API as well.

Could you please clarify what you would expect this reducer to do?  I feel like relying on the follow-up LIST_EXPERIMENT_API action to update the state is fine since some things that are not captured by the createExperimentApi call could change in the future (e.g. default artifact location, additional field when creating an experiment...).

ggliem committed:
Update UUID handling in CreateExperimentModal

Zangr reviewed (commented):
> I could not reproduce the issue locally, however, it was likely caused by the mentioned issue since that did not guarantee the order in which the promises were resolved. That being said, I do not think that the suggested code works since the state of `createExperimentApi` promise has to be shared through the promise chain. I pushed a change - could you please check whether you still get the error?

I got your point! However, it's a bit awkward to have to pass down using Promise.all. What do you think about the following implementation with async/await? Also, actually we are not using request ids in this component so we can actually remove them for simplicity.
```
  handleCreateExperiment = async (values) => {
    // get values of input fields
    const experimentName = values[EXP_NAME_FIELD];
    const artifactLocation = values[ARTIFACT_LOCATION];
    const response = await this.props.createExperimentApi(experimentName, artifactLocation);
    await this.props.listExperimentsApi();
    const { value: { experiment_id: newExperimentId } } = response;
    if (newExperimentId) {
      this.props.history.push(Routes.getExperimentPageRoute(newExperimentId));
    }
  };
```
To use the above syntax, you may want to update .eslintrc.js with the following:
```
    'space-before-function-paren': [2, {
      'anonymous': 'always',
      'named': 'never',
      'asyncArrow': 'always',
    }],
```

> Could you please clarify what you would expect this reducer to do? I feel like relying on the follow-up LIST_EXPERIMENT_API action to update the state is fine since some things that are not captured by the createExperimentApi call could change in the future (e.g. default artifact location, additional field when creating an experiment...).

You are right. Too bad we don't have the complete experiment metadata in the response. We do need the listing for now. No change is needed for the reducer.

ggliem committed:
Update experiment creation to use async/await pattern

ggliem commented:
> I got your point! However, it's a bit awkward to have to pass down using Promise.all. What do you think about the following implementation with async/await? Also, actually we are not using request ids in this component so we can actually remove them for simplicity.

I agree, I initially also looked at the async/await pattern but stuck to the Promise.all implementation since async/await was not used elsewhere in the application. It looks much cleaner now, thanks for the feedback.

ggliem commented:
Python tests seem to be failing again for the last commits

harupy commented:
The CI fails when installing`fbprophet`.
It seems `fbprophet` dropped python 2 support in the latest version (0.6).

Related issue:
https://github.com/facebook/prophet/issues/1339

fbprophet release history:
https://pypi.org/project/fbprophet/0.6/#history

Zangr reviewed (approved):
LGTM Thank you @ggliem, this is a huge UX improvement for experiment CRUD! 
Ready to merge after we did the python build fix.
